### PR TITLE
Reduce the amount of state we pull from the DB

### PR DESCRIPTION
**Pull request by erikjohnston**
Most of these changes are converting calls to the `StateHandler`  for current state to hitting the store instead, which avoids needless state resolution.

Reviewable commit-by-commit.

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -353,15 +362,11 @@ async def _maybe_backfill_inner(
...
 
-        curr_state = await self.state_handler.get_current_state(room_id)
-
-        curr_domains = get_domains_from_state(curr_state)
-
-        likely_domains = [
-            domain for domain, depth in curr_domains if domain != self.server_name
-        ]
+        users_in_room = await self.store.get_users_in_room(room_id)
+        likely_domains = {get_domain_from_id(u) for u in users_in_room}
```
Comment (erikjohnston) [inline synapse/handlers/federation.py:366]: Actually, `get_domains_from_state` was doing more here and ordering the servers by oldest user first.

**Comment by erikjohnston** [comment]
So the test failure is because the `current_state_events` table isn't protected from partial state, unlike the state handler.

I think we need to add 

```python
if await_full_state:
    await self._partial_state_events_tracker.await_full_state(event_ids)
```

when we interact with the current state events table?

File: synapse/api/auth.py [hunk 2]
```diff
@@ -99,29 +98,28 @@ async def check_user_in_room(
...
 
-        if member:
-            membership = member.membership
+        (
+            membership,
+            member_event_id,
+        ) = await self.store.get_local_current_membership_for_user_in_room(
+            user_id=user_id,
+            room_id=room_id,
+        )
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/api/auth.py:109]: Woops, done!

**Comment by richvdh** [comment]
@erikjohnston is this ready for review? What are we doing about the failing test?

**Comment by erikjohnston** [comment]
Oh, err, actually I guess there is no point going through all this until we've fixed the tests. I mean, it is basically ready and usable in non-partial join world, but lets not regress partial joins for this.

**Review by H-Shay — APPROVED** [review]
This appears sane to me.

**Pull request merged**